### DISCUSSION TITLE
is it possible to take a pre built tree and convert it to a maze

### DISCUSSION THREAD

**Discussion by gary-programing** [john-science/mazelib#131] [General]
basically i'm wondering if it would be possible to take a pre built tree with a set number of nodes and convert it to a maze of a certain size using code. i can do it using paper i'm just unsure how i would start to program this
![maze and tree example](https://user-images.githubusercontent.com/61935124/160493494-30b64fd9-73dc-4fb0-baa4-463c003ddade.png)

**Comment by john-science** [comment]
That's a fun idea.

Mazelib certainly doesn't support that workflow out of the box. Nearly every maze-generating algorithm IS based on graph theory, so I bet it's not far off. But it's not supported Yet.

One major difference between the mazes modeled in Mazelib and what you have there is that where you have point `1` connected to point `2` and point `10`, Mazelib would have `1` connected to `10` and that little empty square to the Left of `2`. And THAT square would only lead down to `2`. Does that make sense?  The main difference in how Mazelib represents a maze is that in Mazelib every position in the maze is part of the graph, where in your graph on the right, the only positions that are graphed are positions where choices are made.

I'm not saying that your way is good or bad, it's just different enough it might be hard to model in Mazelib, is all.

**Reply by gary-programing** [reply]
thanks yes it does make sense. this is for my honours project where i'm trying to generate maximally complex mazes and the key factor that we decided on was the number of choices so my complexity score basically became the number of nodes in that tree.

 so this was kind of one of the ways i thought about it where basically the only nodes that needed representing were choices or dead ends and the rest would basically be just edges used to travel to them.

 i choice this way as it would be relitivly easy to find that max complexity by scaling up the number of nodes until it was no longer possible to fit that tree into a maze of a certain size. 

the problem is i'm struggiling to find a way to do what i want to do. another way i thought about was implementing something into the solver that counted the number of choices to be made and then gave that as the complexity score the problem with this one being it's completly random and i don't know how to generate a maze where i can increase the number of choices in the maze every run through to limit the number of mazes i generate but also to know when i've reached that maximimum score.

**Reply by john-science** [reply]
Well, if it's for a project, I probably shouldn't help too much.

But do you about the [Monte Carlo Method](https://en.wikipedia.org/wiki/Monte_Carlo_method)? It is a computational tool used most often in scientific disciplines.

And there is a Monte-Carlo tool built into Mazelib that you could use:

https://github.com/john-science/mazelib/blob/main/docs/API.md#advanced-controlling-maze-difficulty

**Reply by gary-programing** [reply]
cheers i appriciate any help that i can get

**Reply by gary-programing** [reply]
so say for example i generate 100 mazes like you have if i implement a similar method to the difficulty for your mazes but bassically analysising the number of choices in a maze or nodes with more than 1 potential path using the monte carlo method i should be able to pick out the most complex one correct?

**Reply by john-science** [reply]
Sure. 

You would just need to `m.solve()` each maze. And then look at each solution and find some way to count the number of turns in each solution, to get back to your definition of complexity.

**Reply by gary-programing** [reply]
cheers this really helped
